compiler/rustc_codegen_llvm/src/builder/gpu_offload.rs

```diff
@@ -0,0 +1,440 @@
+    }
+
+    // Step 1)
+    unsafe { llvm::LLVMRustPositionBefore(builder.llbuilder, kernel_call) };
+    builder.memset(
+        tgt_bin_desc_alloca,
+        cx.get_const_i8(0),
+        cx.get_const_i64(32),
+        Align::from_bytes(8).unwrap(),
+    );
+
+    let mapper_fn_ty = cx.type_func(&[cx.type_ptr()], cx.type_void());
+    let register_lib_decl = declare_offload_fn(&cx, "__tgt_register_lib", mapper_fn_ty);
+    let unregister_lib_decl = declare_offload_fn(&cx, "__tgt_unregister_lib", mapper_fn_ty);
+    let init_ty = cx.type_func(&[], cx.type_void());
+    let init_rtls_decl = declare_offload_fn(cx, "__tgt_init_all_rtls", init_ty);
+
+    // call void @__tgt_register_lib(ptr noundef %6)
+    builder.call(mapper_fn_ty, register_lib_decl, &[tgt_bin_desc_alloca], None);
+    // call void @__tgt_init_all_rtls()
+    builder.call(init_ty, init_rtls_decl, &[], None);
+
+    for i in 0..num_args {
+        let idx = cx.get_const_i32(i);
+        let gep1 = builder.inbounds_gep(ty, a1, &[i32_0, idx]);
+        builder.store(vals[i as usize], gep1, Align::EIGHT);
+        let gep2 = builder.inbounds_gep(ty, a2, &[i32_0, idx]);
+        builder.store(geps[i as usize], gep2, Align::EIGHT);
+        let gep3 = builder.inbounds_gep(ty2, a4, &[i32_0, idx]);
+        // As mentioned above, we don't use Rust type informatino yet. So for now we will just
+        // assume that we have 1024 bytes, 256 f32 values.
+        // FIXME(offload): write an offload frontend and handle arbitrary types.
+        builder.store(cx.get_const_i64(1024), gep3, Align::EIGHT);
+    }
+
+    // Step 2)
+    let gep1 = builder.inbounds_gep(ty, a1, &[i32_0, i32_0]);
+    let gep2 = builder.inbounds_gep(ty, a2, &[i32_0, i32_0]);
+    let gep3 = builder.inbounds_gep(ty2, a4, &[i32_0, i32_0]);
+
+    let nullptr = cx.const_null(cx.type_ptr());
+    let o_type = o_types[0];
+    let s_ident_t = generate_at_one(&cx);
+    let args = vec![
+        s_ident_t,
+        cx.get_const_i64(u64::MAX),
+        cx.get_const_i32(num_args),
+        gep1,
+        gep2,
+        gep3,
+        o_type,
+        nullptr,
+        nullptr,
+    ];
+    builder.call(fn_ty, begin_mapper_decl, &args, None);
+
+    // Step 3)
+    // Here we will add code for the actual kernel launches in a follow-up PR.
+    // FIXME(offload): launch kernels
+
+    // Step 4)
+    unsafe { llvm::LLVMRustPositionAfter(builder.llbuilder, kernel_call) };
+
+    let gep1 = builder.inbounds_gep(ty, a1, &[i32_0, i32_0]);
+    let gep2 = builder.inbounds_gep(ty, a2, &[i32_0, i32_0]);
+    let gep3 = builder.inbounds_gep(ty2, a4, &[i32_0, i32_0]);
+
+    let nullptr = cx.const_null(cx.type_ptr());
+    let o_type = o_types[0];
+    let args = vec![
+        s_ident_t,
+        cx.get_const_i64(u64::MAX),
+        cx.get_const_i32(num_args),
+        gep1,
+        gep2,
+        gep3,
+        o_type,
+        nullptr,
+        nullptr,
+    ];
+    builder.call(fn_ty, end_mapper_decl, &args, None);
```

Comment:
Since almost the exact same call is gonna be happening thrice in the future, pull this into a function that deduplicates as much as possible

Reply:
I split it up into two functions, since I expect that the gep indexing might see changes when we get to more interesting arguments.